presto-hive/src/main/java/com/facebook/presto/hive/clustering/MortonCode.java

```diff
@@ -0,0 +1,136 @@
+                );
+            }
+
+            PrimitiveTypeInfo pTypeInfo = (PrimitiveTypeInfo) typeInfo;
+            PrimitiveObjectInspector.PrimitiveCategory pCategory = pTypeInfo.getPrimitiveCategory();
+
+            if (pCategory == PrimitiveObjectInspector.PrimitiveCategory.SHORT)
+                return;
+            if (pCategory != PrimitiveObjectInspector.PrimitiveCategory.INT)
+                return;
+            if (pCategory != PrimitiveObjectInspector.PrimitiveCategory.LONG)
+                return;
+
+            throw new UnsupportedOperationException(
+                    "Clustering supports only integer data types!"
+            );
+        }
+    }
+
+    private static List<Integer> getIntervalIndices(
+            List<String> columnNames,
+            Map<String, Interval> columnIntervals,
+            Page page,
+            int position) {
+        List<Integer> indices = new ArrayList<>();
+        for (int i = 0; i < columnNames.size(); ++i) {
+            String columnName = columnNames.get(i);
+            Interval intervals = columnIntervals.get(columnName);
+            int index = getInvervalIndex(intervals, i, page, position);
+            indices.add(index);
+        }
+        return indices;
+    }
+
+    private static int getInvervalIndex(Interval intervals, int idx, Page page, int position)
+    {
+        List<Object> splittingValues = intervals.getSplittingValues();
+        HiveType columnType = intervals.getColumnType();
+
+        int i = 0;
+        for (; i < splittingValues.size(); ++i) {
+            Object splitValue = splittingValues.get(i);
+            if (compare(splitValue, columnType, idx, page, position)) {
+                return i;
+            }
+        }
+        return i;
+    }
+
+    static boolean compare(Object splittingValue, HiveType columnType, int idx, Page page, int position)
+            throws IllegalArgumentException
+    {
+        if (columnType == HiveType.HIVE_INT ||
+                columnType == HiveType.HIVE_LONG ||
+                columnType == HiveType.HIVE_SHORT
+        ) {
+            return (Long) splittingValue >= page.getBlock(idx).getLong(position);
+        } else if (columnType == HiveType.HIVE_DOUBLE ||
+                columnType == HiveType.HIVE_FLOAT
+        ) {
+            // How to get double value?
+            return (Double) splittingValue >= page.getBlock(idx).getLong(position);
+        } else if (columnType == HiveType.HIVE_STRING){
+            // How to get string value?
+            int r = ((String) splittingValue).compareTo(
+                    (String.valueOf(page.getBlock(idx).getLong(position))));
+            return r >= 0;
+        } else {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "Hive clustering does not support column type: %s",
+                            columnType.toString()
+                    )
+            );
+        }
```

Comment:
There are helpers to do the type/value conversion: `nativeValueToBlock` and `blockToNativeValue`. Let's maybe try to use `Type` instead of `HiveType`